src/test/java/org/jenkinsci/remoting/engine/JnlpProtocolHandlerTest.java

```diff
@@ -110,594 +113,197 @@
-                .connect(westChannel.socket(), westProps, new JnlpConnectionStateListener() {
-                    @Override
-                    public void afterProperties(@NonNull JnlpConnectionState event) {
-                        event.approve();
-                    }
-
-                    @Override
-                    public void beforeChannel(@NonNull JnlpConnectionState event) {
-                        event.getChannelBuilder().withMode(Channel.Mode.BINARY);
-                    }
-
-                    @Override
-                    public void afterChannel(@NonNull JnlpConnectionState event) {
-                    }
-                });
-        ByteBuffer len = ByteBuffer.wrap(new byte[2]);
-        while (len.hasRemaining()) {
-            eastChannel.read(len);
-        }
-        byte[] bytes = new byte[((len.get(0) << 8) & 0xff00) + (len.get(1) & 0xff)];
-        ByteBuffer content = ByteBuffer.wrap(bytes);
-        while (content.hasRemaining()) {
-            eastChannel.read(content);
-        }
-        assertThat(new String(bytes, StandardCharsets.UTF_8), is("Protocol:" + factory.toString()));
-        Future<Channel> eastChan = eastProto
-                .handle(eastChannel.socket(), new HashMap<String, String>(), new JnlpConnectionStateListener() {
-                    @Override
-                    public void afterProperties(@NonNull JnlpConnectionState event) {
-                        event.approve();
-                    }
-
-                    @Override
-                    public void beforeChannel(@NonNull JnlpConnectionState event) {
-                        event.getChannelBuilder().withMode(Channel.Mode.NEGOTIATE);
-                    }
-
-                    @Override
-                    public void afterChannel(@NonNull JnlpConnectionState event) {
-                    }
-                });
-        eastRemoting = eastChan.get(10, TimeUnit.SECONDS);
-        assertThat(eastRemoting, notNullValue());
-        eastRemoting.callAsync(new TestCallable());
-        westRemoting = westChan.get(10, TimeUnit.SECONDS);
-        assertThat(westRemoting, notNullValue());
+        JnlpProtocolHandler<? extends JnlpConnectionState> serverProtocolHandler = createServerProtocolHandler(factory, useNioHubServer, SECRET_KEY, true);
+        JnlpProtocolHandler<? extends JnlpConnectionState> clientProtocolHandler = createClientProtocolHandler(factory, useNioHubClient);
+        HashMap<String, String> clientProps = createClientProperties(factory, SECRET_KEY);
+        Future<Channel> clientChannelFuture = createChannelConnector(clientSocketChannel, clientProtocolHandler, clientProps, APPROVING_STATE_CONSUMER);
+        readAndCheckProtocol(factory);
+        Future<Channel> serverChannelFuture = createChannelHandler(serverSocketChannel, serverProtocolHandler, new HashMap<>(), APPROVING_STATE_CONSUMER);
+        serverRemotingChannel = serverChannelFuture.get(10, TimeUnit.SECONDS);
```

Comment:
It may be overkill in this scenario, but I like to use asynchronous-aware libraries when testing async code. For Java, I found [awaitility](https://github.com/awaitility/awaitility) which looks interesting. This type of pattern is a lot more common in Scala since they've had a `Future` class far longer than Java's actual equivalent.

Reply:
That looks like an interesting testing library. I think there's enough change already in this PR that I'd rather not introduce that one at this point.